File: connector/src/main/java/org/geysermc/connector/entity/BoatEntity.java
/*
 * Copyright (c) 2019-2021 GeyserMC. http://geysermc.org
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * @author GeyserMC
 * @link https://github.com/GeyserMC/Geyser
 */

package org.geysermc.connector.entity;

import com.github.steveice10.mc.protocol.data.game.entity.metadata.EntityMetadata;
import com.nukkitx.math.vector.Vector3f;
import com.nukkitx.protocol.bedrock.data.entity.EntityData;
import com.nukkitx.protocol.bedrock.packet.AnimatePacket;
import com.nukkitx.protocol.bedrock.packet.MoveEntityAbsolutePacket;
import org.geysermc.connector.entity.type.EntityType;
import org.geysermc.connector.network.session.GeyserSession;

import java.util.concurrent.TimeUnit;

public class BoatEntity extends Entity {

    /**
     * Required when IS_BUOYANT is sent in order for boats to work in the water. <br>
     *
     * Taken from BDS 1.16.200, with the modification of <code>simulate_waves</code> since Java doesn't bob the boat up and down
     * like Bedrock.
     */
    private static final String BUOYANCY_DATA = "{\"apply_gravity\":true,\"base_buoyancy\":1.0,\"big_wave_probability\":0.02999999932944775," +
            "\"big_wave_speed\":10.0,\"drag_down_on_buoyancy_removed\":0.0,\"liquid_blocks\":[\"minecraft:water\"," +
            "\"minecraft:flowing_water\"],\"simulate_waves\":false}";

    private boolean isPaddlingLeft;
    private float paddleTimeLeft;
    private boolean isPaddlingRight;
    private float paddleTimeRight;

    // Looks too fast and too choppy with 0.1f, which is how I believe the Microsoftian client handles it
    private final float ROWING_SPEED = 0.05f;

    public BoatEntity(long entityId, long geyserId, EntityType entityType, Vector3f position, Vector3f motion, Vector3f rotation) {
        super(entityId, geyserId, entityType, position.add(0d, entityType.getOffset(), 0d), motion, rotation.add(90, 0, 90));

        // Required to be able to move on land 1.16.200+ or apply gravity not in the water 1.16.100+
        metadata.put(EntityData.IS_BUOYANT, (byte) 1);
        metadata.put(EntityData.BUOYANCY_DATA, BUOYANCY_DATA);
    }

    @Override
    public void moveAbsolute(GeyserSession session, Vector3f position, Vector3f rotation, boolean isOnGround, boolean teleported) {
        // We don't include the rotation (y) as it causes the boat to appear sideways
        setPosition(position.add(0d, this.entityType.getOffset(), 0d));
        setRotation(Vector3f.from(rotation.getX() + 90, 0, rotation.getX() + 90));
        setOnGround(isOnGround);

        MoveEntityAbsolutePacket moveEntityPacket = new MoveEntityAbsolutePacket();
        moveEntityPacket.setRuntimeEntityId(geyserId);
        // Minimal glitching when ServerVehicleMovePacket is sent
        moveEntityPacket.setPosition(session.getRidingVehicleEntity() == this ? position.up(EntityType.PLAYER.getOffset() - this.entityType.getOffset()) : this.position);
        moveEntityPacket.setRotation(getBedrockRotation());
        moveEntityPacket.setOnGround(isOnGround);
        moveEntityPacket.setTeleported(teleported);

        session.sendUpstreamPacket(moveEntityPacket);
    }

    @Override
    public void moveRelative(GeyserSession session, double relX, double relY, double relZ, Vector3f rotation, boolean isOnGround) {
        super.moveRelative(session, relX, relY, relZ, Vector3f.from(rotation.getX(), 0, rotation.getX()), isOnGround);
    }

    @Override
    public void updatePositionAndRotation(GeyserSession session, double moveX, double moveY, double moveZ, float yaw, float pitch, boolean isOnGround) {
        moveRelative(session, moveX, moveY, moveZ, yaw + 90, pitch, isOnGround);
    }

    @Override
    public void updateRotation(GeyserSession session, float yaw, float pitch, boolean isOnGround) {
        moveRelative(session, 0, 0, 0, Vector3f.from(yaw + 90, 0, 0), isOnGround);
    }

    @Override
    public void updateBedrockMetadata(EntityMetadata entityMetadata, GeyserSession session) {
        // Time since last hit
        if (entityMetadata.getId() == 8) {
            metadata.put(EntityData.HURT_TIME, entityMetadata.getValue());
        }

        // Rocking direction
        if (entityMetadata.getId() == 9) {
            metadata.put(EntityData.HURT_DIRECTION, entityMetadata.getValue());
        }

        // 'Health' in Bedrock, damage taken in Java
        if (entityMetadata.getId() == 10) {
            // Not exactly health but it makes motion in Bedrock
            metadata.put(EntityData.HEALTH, 40 - ((int) (float) entityMetadata.getValue()));
        }

        if (entityMetadata.getId() == 11) {
            metadata.put(EntityData.VARIANT, entityMetadata.getValue());
        } else if (entityMetadata.getId() == 12) {
            isPaddlingLeft = (boolean) entityMetadata.getValue();
            if (isPaddlingLeft) {
                // Java sends simply "true" and "false" (is_paddling_left), Bedrock keeps sending packets as you're rowing
                // This is an asynchronous method that emulates Bedrock rowing until "false" is sent.
                paddleTimeLeft = 0f;
                if (!this.passengers.isEmpty()) {
                    // Get the entity by the first stored passenger and convey motion in this manner
                    Entity entity = session.getEntityCache().getEntityByJavaId(this.passengers.iterator().nextLong());
                    if (entity != null) {
                        session.getConnector().getGeneralThreadPool().execute(() ->
                                updateLeftPaddle(session, entity)
                        );
                    }
                }
            } else {
                // Indicate that the row position should be reset
                metadata.put(EntityData.ROW_TIME_LEFT, 0.0f);
            }
        }
        else if (entityMetadata.getId() == 13) {
            isPaddlingRight = (boolean) entityMetadata.getValue();
            if (isPaddlingRight) {
                paddleTimeRight = 0f;
                if (!this.passengers.isEmpty()) {
                    Entity entity = session.getEntityCache().getEntityByJavaId(this.passengers.iterator().nextLong());
                    if (entity != null) {
                        session.getConnector().getGeneralThreadPool().execute(() ->
                                updateRightPaddle(session, entity)
                        );
                    }
                }
            } else {
                metadata.put(EntityData.ROW_TIME_RIGHT, 0.0f);
            }
        } else if (entityMetadata.getId() == 14) {
            // Possibly - I don't think this does anything?
            metadata.put(EntityData.BOAT_BUBBLE_TIME, entityMetadata.getValue());
        }

        super.updateBedrockMetadata(entityMetadata, session);
    }

    @Override
    public void updateBedrockMetadata(GeyserSession session) {
        super.updateBedrockMetadata(session);

        // As these indicate to reset rowing, remove them until it is time to send them out again.
        metadata.remove(EntityData.ROW_TIME_LEFT);
        metadata.remove(EntityData.ROW_TIME_RIGHT);
    }

    private void updateLeftPaddle(GeyserSession session, Entity rower) {
        if (isPaddlingLeft) {
            paddleTimeLeft += ROWING_SPEED;
            sendAnimationPacket(session, rower, AnimatePacket.Action.ROW_LEFT, paddleTimeLeft);

            session.getConnector().getGeneralThreadPool().schedule(() ->
                    updateLeftPaddle(session, rower),
                    100,
                    TimeUnit.MILLISECONDS
            );
        }
    }

    private void updateRightPaddle(GeyserSession session, Entity rower) {
        if (isPaddlingRight) {
            paddleTimeRight += ROWING_SPEED;
            sendAnimationPacket(session, rower, AnimatePacket.Action.ROW_RIGHT, paddleTimeRight);

            session.getConnector().getGeneralThreadPool().schedule(() ->
                            updateRightPaddle(session, rower),
                    100,
                    TimeUnit.MILLISECONDS
            );
        }
    }

    private void sendAnimationPacket(GeyserSession session, Entity rower, AnimatePacket.Action action, float rowTime) {
        AnimatePacket packet = new AnimatePacket();
        packet.setRuntimeEntityId(rower.getGeyserId());
        packet.setAction(action);
        packet.setRowingTime(rowTime);
        session.sendUpstreamPacket(packet);
    }
}

File: core/src/main/java/org/geysermc/geyser/entity/type/BoatEntity.cpp
/*
 * Copyright (c) 2019-2022 GeyserMC. http://geysermc.org
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * @author GeyserMC
 * @link https://github.com/GeyserMC/Geyser
 */

package org.geysermc.geyser.entity.type;

#include "lombok.Getter"
#include "org.cloudburstmc.math.vector.Vector2f"
#include "org.cloudburstmc.math.vector.Vector3f"
#include "org.cloudburstmc.protocol.bedrock.data.entity.EntityDataTypes"
#include "org.cloudburstmc.protocol.bedrock.data.entity.EntityFlag"
#include "org.cloudburstmc.protocol.bedrock.packet.MoveEntityAbsolutePacket"
#include "org.geysermc.geyser.entity.spawn.EntitySpawnContext"
#include "org.geysermc.geyser.entity.vehicle.BoatVehicleComponent"
#include "org.geysermc.geyser.entity.vehicle.ClientVehicle"
#include "org.geysermc.geyser.entity.vehicle.VehicleComponent"
#include "org.geysermc.geyser.util.InteractionResult"
#include "org.geysermc.geyser.util.InteractiveTag"
#include "org.geysermc.mcprotocollib.protocol.data.game.entity.metadata.type.BooleanEntityMetadata"
#include "org.geysermc.mcprotocollib.protocol.data.game.entity.player.Hand"
#include "org.geysermc.mcprotocollib.protocol.packet.ingame.serverbound.level.ServerboundPaddleBoatPacket"

public class BoatEntity extends Entity implements Tickable, Leashable, ClientVehicle {


    private static final std::string BUOYANCY_DATA = "{\"apply_gravity\":true,\"base_buoyancy\":1.0,\"big_wave_probability\":0.02999999932944775," +
            "\"big_wave_speed\":10.0,\"drag_down_on_buoyancy_removed\":0.0,\"liquid_blocks\":[\"minecraft:water\"," +
            "\"minecraft:flowing_water\"],\"simulate_waves\":false}";

    private final BoatVehicleComponent vehicleComponent = new BoatVehicleComponent(this, 0);

    private bool isPaddlingLeft;
    private float paddleTimeLeft;
    private bool isPaddlingRight;
    private float paddleTimeRight;
    private bool doTick;


    @Getter
    protected final BoatVariant variant;

    private long leashHolderBedrockId = -1;


    private final float ROWING_SPEED = 0.04f;

    public BoatEntity(EntitySpawnContext context, BoatVariant variant) {
        super(context);

        setYaw(yaw + 90);
        setPitch(0);
        setHeadYaw(headYaw + 90);
        this.variant = variant;

        dirtyMetadata.put(EntityDataTypes.VARIANT, variant.ordinal());


        dirtyMetadata.put(EntityDataTypes.IS_BUOYANT, true);
        dirtyMetadata.put(EntityDataTypes.BUOYANCY_DATA, BUOYANCY_DATA);;
    }

    override protected void initializeMetadata() {
        super.initializeMetadata();

        setFlag(EntityFlag.COLLIDABLE, true);
    }

    override public void moveAbsoluteRaw(Vector3f position, float yaw, float pitch, float headYaw, bool isOnGround, bool teleported) {

        setPosition(position);
        setYaw(yaw + 90);
        setHeadYaw(yaw + 90);
        setOnGround(isOnGround);

        MoveEntityAbsolutePacket moveEntityPacket = new MoveEntityAbsolutePacket();
        moveEntityPacket.setRuntimeEntityId(geyserId);
        moveEntityPacket.setPosition(bedrockPosition());
        moveEntityPacket.setRotation(bedrockRotation());
        moveEntityPacket.setOnGround(isOnGround);
        moveEntityPacket.setTeleported(teleported);

        session.sendUpstreamPacket(moveEntityPacket);
    }

    override public void moveRelativeRaw(double relX, double relY, double relZ, float yaw, float pitch, float headYaw, bool isOnGround) {
        super.moveRelativeRaw(relX, relY, relZ, yaw, 0, yaw, isOnGround);
    }

    override public void updatePositionAndRotation(double moveX, double moveY, double moveZ, float yaw, float pitch, bool isOnGround) {
        moveRelative(moveX, moveY, moveZ, yaw + 90, 0, 0, isOnGround);
    }

    override public void updateRotation(float yaw, float pitch, bool isOnGround) {
        moveRelativeRaw(0, 0, 0, yaw + 90, 0, 0, isOnGround);
    }

    public void setPaddlingLeft(BooleanEntityMetadata entityMetadata) {
        isPaddlingLeft = entityMetadata.getPrimitiveValue();
        if (!isPaddlingLeft) {
            paddleTimeLeft = 0.0f;
            dirtyMetadata.put(EntityDataTypes.ROW_TIME_LEFT, 0.0f);
        }
    }

    public void setPaddlingRight(BooleanEntityMetadata entityMetadata) {
        isPaddlingRight = entityMetadata.getPrimitiveValue();
        if (!isPaddlingRight) {
            paddleTimeRight = 0.0f;
            dirtyMetadata.put(EntityDataTypes.ROW_TIME_RIGHT, 0.0f);
        }
    }

    override public void setLeashHolderBedrockId(long bedrockId) {
        this.leashHolderBedrockId = bedrockId;
        dirtyMetadata.put(EntityDataTypes.LEASH_HOLDER, bedrockId);
    }

    override protected InteractiveTag testInteraction(Hand hand) {
        InteractiveTag tag = super.testInteraction(hand);
        if (tag != InteractiveTag.NONE) {
            return tag;
        }
        if (session.isSneaking()) {
            return InteractiveTag.NONE;
        } else if (passengers.size() < 2) {
            return InteractiveTag.BOARD_BOAT;
        } else {
            return InteractiveTag.NONE;
        }
    }

    override public InteractionResult interact(Hand hand) {
        InteractionResult result = super.interact(hand);
        if (result != InteractionResult.PASS) {
            return result;
        }
        if (session.isSneaking()) {
            return InteractionResult.PASS;
        } else {

            return InteractionResult.SUCCESS;
        }
    }

    override public void tick() {

        if (session.getPlayerEntity().getVehicle() == this) {

            ServerboundPaddleBoatPacket steerPacket = new ServerboundPaddleBoatPacket(session.isSteeringLeft(), session.isSteeringRight());
            session.sendDownstreamGamePacket(steerPacket);


            if (session.isInClientPredictedVehicle()) {
                return;
            }
        }

        doTick = !doTick; // Run every other tick
        if (!doTick || passengers.isEmpty()) {
            return;
        }

        Entity rower = passengers.get(0);
        if (rower == null) {
            return;
        }

        if (isPaddlingLeft) {
            paddleTimeLeft += ROWING_SPEED;
            dirtyMetadata.put(EntityDataTypes.ROW_TIME_LEFT, paddleTimeLeft);
        }
        if (isPaddlingRight) {
            paddleTimeRight += ROWING_SPEED;
            dirtyMetadata.put(EntityDataTypes.ROW_TIME_RIGHT, paddleTimeRight);
        }

        if (isPaddlingLeft || isPaddlingRight) {
            updateBedrockMetadata();
        }
    }

    override public long leashHolderBedrockId() {
        return this.leashHolderBedrockId;
    }

    override public VehicleComponent<?> getVehicleComponent() {
        return this.vehicleComponent;
    }

    override public Vector3f getRiddenInput(Vector2f input) {
        return Vector3f.ZERO;
    }

    override public float getVehicleSpeed() {
        return 0;
    }

    override public bool shouldSimulateMovement() {
        return !session.isInClientPredictedVehicle() && !passengers.isEmpty() && this.session.getPlayerEntity() == passengers.get(0);
    }


    public enum BoatVariant {
        OAK,
        SPRUCE,
        BIRCH,
        JUNGLE,
        ACACIA,
        DARK_OAK,
        MANGROVE,
        BAMBOO,
        CHERRY,
        PALE_OAK;

        BoatVariant() {}
    }
}
